File: backend/api/credits.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from typing import Optional
from datetime import datetime

from database import get_db
from services.credit_service import CreditService
from schemas.credits import (
    CreditBalance, CreditTransactionHistory, CreditTransaction,
    AddCreditsRequest, RemoveCreditsRequest
)
from api.auth import get_current_user
from core.exceptions import (
    InsufficientCreditsError, CreditServiceError, UserNotFoundError,
    InvalidAmountError, TransactionError
)

router = APIRouter(prefix="/credits", tags=["credits"])

@router.get("/balance", response_model=CreditBalance)
async def get_credit_balance(
    current_user = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Get current credit balance for the authenticated user"""
    try:
        credit_service = CreditService(db)
        return credit_service.get_user_balance(current_user.id)
    except UserNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail="Failed to get credit balance")

@router.get("/summary")
async def get_credit_summary(
    current_user = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Get comprehensive credit summary for the authenticated user"""
    try:
        credit_service = CreditService(db)
        return credit_service.get_credit_summary(current_user.id)
    except UserNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail="Failed to get credit summary")

@router.get("/transactions", response_model=CreditTransactionHistory)
async def get_transaction_history(
    page: int = Query(1, ge=1, description="Page number"),
    page_size: int = Query(50, ge=1, le=100, description="Page size"),
    transaction_type: Optional[str] = Query(None, description="Filter by transaction type"),
    start_date: Optional[datetime] = Query(None, description="Filter transactions from this date"),
    end_date: Optional[datetime] = Query(None, description="Filter transactions until this date"),
    current_user = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Get paginated transaction history for the authenticated user"""
    try:
        credit_service = CreditService(db)
        return credit_service.get_transaction_history(
            user_id=current_user.id,
            page=page,
            page_size=page_size,
            transaction_type=transaction_type,
            start_date=start_date,
            end_date=end_date
        )
    except UserNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail="Failed to get transaction history")

@router.post("/purchase/validate")
async def validate_credit_purchase(
    amount: int = Query(..., gt=0, description="Credits to purchase"),
    current_user = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Validate if user can purchase credits and get estimated cost"""
    try:
        credit_service = CreditService(db)
        return credit_service.validate_credit_purchase(current_user.id, amount)
    except UserNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail="Failed to validate credit purchase")

@router.post("/purchase", response_model=CreditTransaction)
async def purchase_credits(
    amount: int = Query(..., gt=0, description="Credits to purchase"),
    description: str = Query(..., description="Purchase description"),
    stripe_payment_intent_id: Optional[str] = Query(None, description="Stripe payment intent ID"),
    current_user = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Purchase credits (requires successful Stripe payment)"""
    try:
        credit_service = CreditService(db)
        return credit_service.add_credits(
            user_id=current_user.id,
            amount=amount,
            description=description,
            transaction_type="purchase",
            stripe_payment_intent_id=stripe_payment_intent_id,
            metadata={"purchase_type": "one_time"}
        )
    except UserNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except CreditServiceError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail="Failed to purchase credits")

@router.get("/rate")
async def get_credit_rate():
    """Get current credit rate (price per credit)"""
    return {
        "credit_rate": 0.10,
        "currency": "USD",
        "description": "Each credit costs $0.10"
    }

@router.get("/packages")
async def get_credit_packages(
    db: Session = Depends(get_db)
):
    """Get available credit subscription packages"""
    # This will be implemented when CreditPackage model is available
    return {
        "message": "Credit packages endpoint - to be implemented",
        "packages": []
    }

File: frontend/services/credits.ts
import { api } from '../lib/https';
import {
  CreditBalance,
  CreditTransactionHistory,
  CreditTransaction,
  CreditPurchaseRequest,
  CreditPurchaseValidation,
  UserCreditSummary,
  CreditRate,
  CreditPackageResponse
} from '@/types';

export class CreditsService {
  private baseUrl = '/api/credits';

  /**
   * Get current credit balance for the authenticated user
   */
  async getBalance(): Promise<CreditBalance> {
    const response = await apiClient.get<CreditBalance>(`${this.baseUrl}/balance`);
    return response.data;
  }

  /**
   * Get comprehensive credit summary for the authenticated user
   */
  async getSummary(): Promise<UserCreditSummary> {
    const response = await apiClient.get<UserCreditSummary>(`${this.baseUrl}/summary`);
    return response.data;
  }

  /**
   * Get paginated transaction history for the authenticated user
   */
  async getTransactionHistory(
    page: number = 1,
    pageSize: number = 50,
    transactionType?: string,
    startDate?: string,
    endDate?: string
  ): Promise<CreditTransactionHistory> {
    const params = new URLSearchParams({
      page: page.toString(),
      page_size: pageSize.toString()
    });

    if (transactionType) params.append('transaction_type', transactionType);
    if (startDate) params.append('start_date', startDate);
    if (endDate) params.append('end_date', endDate);

    const response = await apiClient.get<CreditTransactionHistory>(
      `${this.baseUrl}/transactions?${params.toString()}`
    );
    return response.data;
  }

  /**
   * Validate if user can purchase credits and get estimated cost
   */
  async validatePurchase(amount: number): Promise<CreditPurchaseValidation> {
    const response = await apiClient.post<CreditPurchaseValidation>(
      `${this.baseUrl}/purchase/validate`,
      null,
      { params: { amount } }
    );
    return response.data;
  }

  /**
   * Purchase credits (requires successful Stripe payment)
   */
  async purchaseCredits(
    amount: number,
    description: string,
    stripePaymentIntentId?: string
  ): Promise<CreditTransaction> {
    const params = new URLSearchParams({
      amount: amount.toString(),
      description
    });

    if (stripePaymentIntentId) {
      params.append('stripe_payment_intent_id', stripePaymentIntentId);
    }

    const response = await apiClient.post<CreditTransaction>(
      `${this.baseUrl}/purchase?${params.toString()}`
    );
    return response.data;
  }

  /**
   * Get current credit rate (price per credit)
   */
  async getCreditRate(): Promise<CreditRate> {
    const response = await apiClient.get<CreditRate>(`${this.baseUrl}/rate`);
    return response.data;
  }

  /**
   * Get available credit subscription packages
   */
  async getCreditPackages(): Promise<CreditPackageResponse> {
    const response = await apiClient.get<CreditPackageResponse>(`${this.baseUrl}/packages`);
    return response.data;
  }
}

// Admin Credits Service
export class AdminCreditsService {
  private baseUrl = '/api/admin/credits';

  /**
   * Add credits to a user account
   */
  async addCredits(
    userId: number,
    amount: number,
    reason: string,
    adminNotes?: string
  ): Promise<any> {
    const response = await apiClient.post(`${this.baseUrl}/add`, {
      user_id: userId,
      amount,
      reason,
      admin_notes: adminNotes
    });
    return response.data;
  }

  /**
   * Remove credits from a user account
   */
  async removeCredits(
    userId: number,
    amount: number,
    reason: string,
    adminNotes?: string
  ): Promise<any> {
    const response = await apiClient.post(`${this.baseUrl}/remove`, {
      user_id: userId,
      amount,
      reason,
      admin_notes: adminNotes
    });
    return response.data;
  }

  /**
   * Pause credit service for a user
   */
  async pauseCreditService(
    userId: number,
    reason: string,
    adminNotes?: string
  ): Promise<any> {
    const response = await apiClient.post(`${this.baseUrl}/pause`, {
      user_id: userId,
      reason,
      admin_notes: adminNotes
    });
    return response.data;
  }

  /**
   * Resume credit service for a user
   */
  async resumeCreditService(
    userId: number,
    adminNotes?: string
  ): Promise<any> {
    const response = await apiClient.post(`${this.baseUrl}/resume`, {
      user_id: userId,
      admin_notes: adminNotes
    });
    return response.data;
  }

  /**
   * Get credit balance for a specific user
   */
  async getUserBalance(userId: number): Promise<CreditBalance> {
    const response = await apiClient.get<CreditBalance>(`${this.baseUrl}/user/${userId}/balance`);
    return response.data;
  }

  /**
   * Get transaction history for a specific user
   */
  async getUserTransactions(
    userId: number,
    page: number = 1,
    pageSize: number = 50,
    transactionType?: string,
    startDate?: string,
    endDate?: string
  ): Promise<CreditTransactionHistory> {
    const params = new URLSearchParams({
      page: page.toString(),
      page_size: pageSize.toString()
    });

    if (transactionType) params.append('transaction_type', transactionType);
    if (startDate) params.append('start_date', startDate);
    if (endDate) params.append('end_date', endDate);

    const response = await apiClient.get<CreditTransactionHistory>(
      `${this.baseUrl}/user/${userId}/transactions?${params.toString()}`
    );
    return response.data;
  }

  /**
   * Get comprehensive credit summary for a specific user
   */
  async getUserSummary(userId: number): Promise<UserCreditSummary> {
    const response = await apiClient.get<UserCreditSummary>(`${this.baseUrl}/user/${userId}/summary`);
    return response.data;
  }

  /**
   * Get system-wide credit summary
   */
  async getSystemSummary(): Promise<any> {
    const response = await apiClient.get(`${this.baseUrl}/system/summary`);
    return response.data;
  }

  /**
   * Get paginated list of users with their credit status
   */
  async getUsersCreditStatus(
    page: number = 1,
    pageSize: number = 50,
    statusFilter?: string
  ): Promise<any> {
    const params = new URLSearchParams({
      page: page.toString(),
      page_size: pageSize.toString()
    });

    if (statusFilter) params.append('status_filter', statusFilter);

    const response = await apiClient.get(`${this.baseUrl}/users/status?${params.toString()}`);
    return response.data;
  }
}

// Disputes Service
export class DisputesService {
  private baseUrl = '/api/disputes';

  /**
   * Submit a new credit dispute
   */
  async submitDispute(disputeData: {
    user_id: number;
    transaction_id?: string;
    reason: string;
    description: string;
    requested_refund?: number;
  }): Promise<any> {
    const response = await apiClient.post(`${this.baseUrl}/submit`, disputeData);
    return response.data;
  }

  /**
   * Get disputes submitted by the current user
   */
  async getMyDisputes(status?: string): Promise<any[]> {
    const params = status ? `?status=${status}` : '';
    const response = await apiClient.get<any[]>(`${this.baseUrl}/my-disputes${params}`);
    return response.data;
  }

  /**
   * Get details of a specific dispute
   */
  async getDispute(disputeId: number): Promise<any> {
    const response = await apiClient.get<any>(`${this.baseUrl}/${disputeId}`);
    return response.data;
  }

  /**
   * Admin: Get dispute queue
   */
  async getDisputeQueue(
    status?: string,
    page: number = 1,
    pageSize: number = 50
  ): Promise<any> {
    const params = new URLSearchParams({
      page: page.toString(),
      page_size: pageSize.toString()
    });

    if (status) params.append('status', status);

    const response = await apiClient.get(`${this.baseUrl}/admin/queue?${params.toString()}`);
    return response.data;
  }

  /**
   * Admin: Review and update dispute status
   */
  async reviewDispute(
    disputeId: number,
    updateData: {
      status?: string;
      resolution?: string;
      resolved_amount?: number;
      admin_notes?: string;
      resolution_notes?: string;
    }
  ): Promise<any> {
    const response = await apiClient.put(`${this.baseUrl}/admin/${disputeId}/review`, updateData);
    return response.data;
  }

  /**
   * Admin: Resolve a dispute
   */
  async resolveDispute(
    disputeId: number,
    resolutionData: {
      resolution: string;
      resolved_amount?: number;
      resolution_notes: string;
      admin_notes?: string;
    }
  ): Promise<any> {
    const response = await apiClient.post(`${this.baseUrl}/admin/${disputeId}/resolve`, resolutionData);
    return response.data;
  }

  /**
   * Admin: Get dispute statistics
   */
  async getDisputeStatistics(): Promise<any> {
    const response = await apiClient.get(`${this.baseUrl}/admin/statistics`);
    return response.data;
  }
}

// Export service instances
export const creditsService = new CreditsService();
export const adminCreditsService = new AdminCreditsService();
export const disputesService = new DisputesService();
